### ISSUE TITLE
Reproduce published results from Yoon Kim 2014

### ISSUE THREAD

**Issue by wpm** [wpm/mycroft#11]
Use this code to reproduce published results for RNNs for text classification. Put the results and references in the README.

Nothing exhaustive, just reproduce something to show that it's working.

**Comment by wpm** [comment]
Maybe the [Stanford sentiment analysis](https://nlp.stanford.edu/sentiment/).

**Comment by wpm** [comment]
This [Keras tutorial](https://blog.keras.io/using-pre-trained-word-embeddings-in-a-keras-model.html) gets 95% accuracy on 20 Newsgroups.

I'm nowhere near this. I must have a bug.

**Comment by wpm** [comment]
I was [confused](https://datascience.stackexchange.com/questions/20451/how-do-you-map-phrase-ids-to-sentence-ids-in-the-stanford-sentiment-analysis-dat) about how what the data set for the Stanford Sentiment Treebank is. Is think the answer is that published results only consider full sentences.

**Comment by wpm** [comment]
Yoon Kim 2014 "[Convolutional neural networks for sentence classification](http://www.aclweb.org/anthology/D14-1181)" gives results for a sequential neural net classifier over several different data sets.

**Comment by wpm** [comment]
* MR:Moviereviewswithonesentenceperre- view. Classification involves detecting posi- tive/negative reviews (Pang and Lee, 2005).3
* SST-1: Stanford Sentiment Treebank—an extension of MR but with train/dev/test splits provided and fine-grained labels (very pos- itive, positive, neutral, negative, very nega- tive), re-labeled by Socher et al. (2013).4
* SST-2: Same as SST-1 but with neutral re- views removed and binary labels.
* Subj: Subjectivity dataset where the task is to classify a sentence as being subjective or objective (Pang and Lee, 2004).
* TREC: TREC question dataset—task in- volves classifying a question into 6 question types (whether the question is about person, location, numeric information, etc.) (Li and Roth, 2002).5
* CR: Customer reviews of various products (cameras, MP3s etc.). Task is to predict pos- itive/negative reviews (Hu and Liu, 2004).6

**Comment by wpm** [comment]
The CR link is 404. (And the paper is written in Microsoft Word. Ick.)

**Comment by wpm** [comment]
Some statistics about the Stanford Sentiment Treebank data set:

| Partition | Longest Text | Vocabulary Size |
| -------- | ------------- | ----------------- |
| Train      | 55                   | 1,864,603            |
| Dev        | 49                   | 20,432                |
| Test        | 56                   | 41,724                 |

Set the sequence length to 60. Maybe increase the vocabulary size to 40,000.

**Comment by wpm** [comment]
To call this done I need
* Loss
* Accuracy
* Training time

for my best run of

* Bag of words
* RNN
* Convolution Net

on the following data sets

* MR
* SST1
* SST2
* SUBJ

Plus the best accuracy I can find in the literature (probably just from the Yoon Kim paper) for each of these data sets.

**Comment by wpm** [comment]
| Model | Data Set | Training Time | Loss | Accuracy |
| ------ | --------- | ------------- | ----- | ---------- |
| Conv   |  SST Coarse |                        |          |                   |
| RNN   |   SST Coarse      |                        |          |                   |
| BOW   |  SST Coarse  |                        |          |                   |
| Paragraph-Vec  | SST Coarse  |                        |          |        48.7           |
| Conv   |  SST Fine |                        |          |                   |
| RNN   |  SST Fine |                        |          |                   |
| BOW   | SST Fine |                        |          |                   |
| CNN-multichannel   | SST Fine |                        |          |       88.1            |
| Conv   | SUBJ |                        |          |                   |
| RNN   | SUBJ  |                        |          |                   |
| BOW   | SUBJ   |                        |          |                   |
| F-Dropout   | SUBJ   |                        |          |         93.6          |

* Paragraph-Vec: Le and Mikolov, 2014
* CNN-multichannel: Yoon Kim, 2014
* F-Dropout: Wang and Manning, 2013

**Comment by wpm** [comment]
Parameter grid directions to explore:
* Batch size (64)
* Dropout (0.25)
* Different RNN units (64, 256)
* Multiple RNN layers
* Vocabulary (200,000)
* Convnet kernel (4, 5)
* Convet filters (50, 150)

**Comment by wpm** [comment]
* Stanford Coarse Conv Full: [Experiment 46](https://www.floydhub.com/billmcn/projects/yoon-kim-2014/46)
* Stanford Fine Conv Full: [Experiment 47](https://www.floydhub.com/billmcn/projects/yoon-kim-2014/47)
